internal/provider/resource_gitlab_project_protected_environment.go

```diff
@@ -0,0 +1,206 @@
+			"project": {
+				Description:  "The ID or full path of the project which the protected environment is created against.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"environment": {
+				Description:  "The name of the environment.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+		// Placeholder attribute till GitLab makes approval count live https://gitlab.com/gitlab-org/gitlab/-/issues/343864
+		// 	"required_approval_count": {
+		// 		Description:  "The number of approvals required to deploy to this environment. This is part of Deployment Approvals, which isn't yet available for use.",
+		// 		Type:         schema.TypeString,
+		// 		ForceNew:     true,
+		// 		Required:     false,
+		// 		ValidateFunc: validation.StringIsNotEmpty,
+		// },
+			"deploy_access_levels": {
+				Description: "Array of access levels allowed to deploy, with each described by a hash.",
+				Type:        schema.TypeList,
+				ForceNew:    true,
+				Required:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"access_level": {
+							Description:  fmt.Sprintf("Levels of access required to deploy to this protected environment. Valid values are %s.", renderValueListForDocs(validProtectedEnvironmentDeploymentLevelNames)),
+							Type:         schema.TypeString,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.StringInSlice(validProtectedEnvironmentDeploymentLevelNames, false),
+						},
+						"access_level_description": {
+							Description: "Readable description of level of access.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"user_id": {
+							Description:  "The ID of the user allowed to deploy to this protected environment.",
+							Type:         schema.TypeInt,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.IntAtLeast(1),
+						},
+						"group_id": {
+							Description:  "The ID of the group allowed to deploy to this protected environment.",
+							Type:         schema.TypeInt,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.IntAtLeast(1),
+						},
+					},
+				},
+			},
+		},
+	}
+})
+
+func resourceGitlabProjectProtectedEnvironmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	deployAccessLevels, err := expandDeployAccessLevels(d.Get("deploy_access_levels").([]interface{}))
+	if err != nil {
+		return diag.Errorf("error expanding deploy_access_levels: %v", err)
+	}
+
+	options := &gitlab.ProtectRepositoryEnvironmentsOptions{
+		Name: gitlab.String(d.Get("environment").(string)),
+		// IDK What needs to go here instead but everything seems to be prepended by ampersands for these
```

Comment:
Have not touched a pointer/reference based language in a long time since my C++ classes so had to brush up on them for this work.  I have removed the comment and now believe the code is good for this assignment